xonsh/teepty.py

```diff
@@ -0,0 +1,234 @@
+"""This implements a psuedo-TTY that tees its output into a Python buffer.
+
+This file was forked from a version distibuted under an MIT license and
+Copyright (c) 2011 Joshua D. Bartlett. 
+See http://sqizit.bartletts.id.au/2011/02/14/pseudo-terminals-in-python/ for 
+more information.
+"""
+import io
+import re
+import os
+import sys
+import tty
+import pty
+import array
+import fcntl
+import select
+import signal
+import termios
+import tempfile
+
+# The following escape codes are xterm codes.
+# See http://rtfm.etla.org/xterm/ctlseq.html for more.
+MODE_NUMS = ('1049', '47', '1047')
+START_ALTERNATE_MODE = frozenset('\x1b[?{0}h'.format(i).encode() for i in MODE_NUMS)
+END_ALTERNATE_MODE = frozenset('\x1b[?{0}l'.format(i).encode() for i in MODE_NUMS)
+ALTERNATE_MODE_FLAGS = tuple(START_ALTERNATE_MODE) + tuple(END_ALTERNATE_MODE)
+
+RE_HIDDEN = re.compile(b'(\001.*?\002)')
```

Comment:
This is entirely pedantic, but if I understand the readline behavior correctly, the regular expression should be `\001[^\001\002]*\002`.